trace-test-application/src/main/java/io/bitrise/trace/testapp/exception/ExceptionActivity.java

```diff
@@ -0,0 +1,127 @@
+ */
+public class ExceptionActivity extends AppCompatActivity {
+
+  /**
+   * The {@link SessionManager} to control session starts and ends.
+   */
+  private final SessionManager sessionManager = ApplicationSessionManager.getInstance();
+
+  @Override
+  protected void onCreate(@Nullable final Bundle savedInstanceState) {
+    super.onCreate(savedInstanceState);
+    setContentView(R.layout.activity_exception);
+    setupView();
+  }
+
+  /**
+   * Sets up the initial state of the view.
+   */
+  private void setupView() {
+    final SwitchCompat sessionSwitch = findViewById(R.id.switch_session);
+    setSessionState(sessionManager.getActiveSession() != null);
+    sessionSwitch.setOnCheckedChangeListener((buttonView, isChecked) -> setSessionState(isChecked));
+
+    final Button exceptionInAppButton = findViewById(R.id.btn_exception_app);
+    final Button exceptionInDependencyButton = findViewById(R.id.btn_exception_dependency);
+    final Button exceptionInDependencyRethrownButton =
+        findViewById(R.id.btn_exception_dependency_rethrow);
+    final Button exceptionInMainThreadButton = findViewById(R.id.btn_exception_main_thread);
+    final Button exceptionInThreadButton = findViewById(R.id.btn_exception_thread);
+
+    exceptionInAppButton.setOnClickListener(v -> throwExceptionInApp());
+    exceptionInDependencyButton.setOnClickListener(v -> throwExceptionInDependency());
+    exceptionInDependencyRethrownButton.setOnClickListener(v -> rethrowExceptionFromDependency());
+    exceptionInMainThreadButton.setOnClickListener(v -> throwExceptionInMainThread());
+    exceptionInThreadButton.setOnClickListener(v -> throwExceptionInThread());
+  }
+
+  /**
+   * Sets the state of the session for the {@link #sessionManager}.
+   *
+   * @param started {@code true} to start the session, {@code false} to stop.
+   */
+  private void setSessionState(final boolean started) {
+    if (started) {
+      sessionManager.startSession();
+    } else {
+      sessionManager.stopSession();
+    }
+  }
+
+  /**
+   * Method for throwing and exception in the applications code.
+   */
+  private void throwExceptionInApp() {
+    throwException("App is not happy!");
+  }
+
+  /**
+   * Method to throw an exception from a dependency's code.
+   */
+  private void throwExceptionInDependency() {
+    throwExceptionInOkHttp();
+  }
+
+  /**
+   * Catches and rethrows an exception.
+   */
+  private void rethrowExceptionFromDependency() {
+    try {
+      throwExceptionInOkHttp();
+    } catch (final NullPointerException npe) {
+      throw new UnsupportedOperationException("Exception caught and rethrown!", npe);
+    }
+  }
+
+  /**
+   * Throws an exception in the main thread. Should be called from the main thread.
+   */
+  private void throwExceptionInMainThread() {
+    throwException("Main thread is not happy!");
+  }
+
+  /**
+   * Creates a new thread and throws an exception.
+   */
+  private void throwExceptionInThread() {
```

Comment:
nit: I'd rename this as `throwExceptionInNewThread` and replace the button text too - it's a bit confusing as you have Main Thread, and then Thread (of which Main Thread is a Thread) - explicitly adding this is a "new" thread would help readability.